### ISSUE TITLE
False positive on noUnusedLocals error

### ISSUE THREAD

**Issue by JesseParsons** [microsoft/TypeScript#54918]
# Bug Report

### 🔎 Search Terms

nounusedlocals false positive

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `noUnusedLocals`

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?noUnusedLocals=true&ts=5.1.6#code/MYGwhgzhAEDCBQBve1XQA4CcCWA3MALgKbQAeAXNBATgHYDmA3CmsAPa3WYCuwBbmABQBKFqmRpJ0AgAtsEAHSloAXmgAiMOuaSAvvDEYc+YtAC2q6ADNutPtg4jDEqalnylC4DLCYAggSCAAzCOmj6ukA)

### 💻 Code

```ts
class C
{
    private x: string;
    constructor()
    {
        this.x = "a";
    }

    private m = function()
    {
        this.x.charAt(0);
    }
}
```

### 🙁 Actual behavior

`x` is line 3 is flagged for `noUnusedLocals` even though it is read in line 11.  If I declare `m` like `private m()`, the error goes away.

### 🙂 Expected behavior

No `noUnusedLocals` violation since the variable is read

**Comment by RyanCavanaugh** [comment]
The `this` in `m` is not `C` (as shown by the implicit any error), so the usage isn't detected.

You should write
```ts
    private m = function(this: C)
    {
        this.x.charAt(0);
    }
```

**Comment by JesseParsons** [comment]
Thanks @RyanCavanaugh is there a recommendation to use `private m()` vs. `private m = function(this: C)` for a private method declaration?  Might be misunderstanding something deeper in how the latter is declared, or perhaps it is simply a personal convention to choose between the two.

**Comment by fatcerberus** [comment]
There is, in fact, a significant difference in runtime behavior: `private m = function...` will put the method on the instance as an enumerable property (class fields).  `private m()` will make it non-enumerable and put it on the prototype, where it's accessible via `super`.  All together I would generally recommend the method syntax unless you have a specific need to define it as a field.
